staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -762,14 +762,35 @@ func shouldDelegateList(opts storage.ListOptions) bool {
 	consistentReadFromStorage := resourceVersion == "" && !(consistentListFromCacheEnabled && requestWatchProgressSupported)
 	// Watch cache doesn't support continuations, so serve them from etcd.
 	hasContinuation := len(pred.Continue) > 0
-	// Serve paginated requests about revision "0" from watch cache to avoid overwhelming etcd.
-	hasLimit := pred.Limit > 0 && resourceVersion != "0"
+	// Serve paginated requests about revision "0" and "" (when consistent read from cache is supported)
+	// from watch cache to avoid overwhelming etcd.
+	consistentListFromCache := resourceVersion == "" && (consistentListFromCacheEnabled && requestWatchProgressSupported)
+	hasLimit := pred.Limit > 0 && !(resourceVersion == "0" || consistentListFromCache)
 	// Watch cache only supports ResourceVersionMatchNotOlderThan (default).
 	unsupportedMatch := match != "" && match != metav1.ResourceVersionMatchNotOlderThan
 
 	return consistentReadFromStorage || hasContinuation || hasLimit || unsupportedMatch
 }
 
+// shouldApplyListLimit determines whether the cacher should
+// apply a limit to an incoming LIST request.
+//
+// note that this function doesn't check RVM nor the Continuation token.
+// these parameters are validated by the shouldDelegateList function.
+//
+// as of today, the limit is applied only to requests that
+// request a consistent snapshot of data from the cache.
+func shouldApplyListLimit(opts storage.ListOptions, requestWatchProgressSupported bool) bool {
+	if opts.Predicate.Limit <= 0 {
+		return false
+	}
+	consistentListFromCacheEnabled := utilfeature.DefaultFeatureGate.Enabled(features.ConsistentListFromCache)
```

Comment:
Following up on the comment about  removing `hasLimit` completely. Wouldn't this signficantly simplify this method too?

The only usecase we want to ignore the limit is the case of `resourceVersion='0' then, so this function would become trivial...

Reply:
yes, removing `hasLimit` will make this function much simpler.